### PR TITLE
Add detect_graspable_poses_pcabase.py and its sample

### PR DESCRIPTION
**Pull request by sshige**
detect_graspable_poses_pcabase.py is a program which publishes posearray where a robot can grasp using point cloud of a target object and PCA algorithm.
Please transform input cloud to a flame which direction is the same as the robot (e.g. /BODY) before using this program.
Parameters of this program are
- direction: from which direction a robot try to grasp a target object. ('x': from the front, 'z': from the top)
- hand_width[m]: How long a robot can spread her/his hand.
- interval_m[m]: Interval between target poses. Please increase this number if the quality of point cloud data is bad.

### CODE REVIEW

**Comment by wkentaro** [comment]
Could you provide the sample image?

Also, why is the below assumption needed for this program?

> Please transform input cloud to a flame which direction is the same as the robot (e.g. /BODY) before using this program.

**Comment by sshige** [comment]
Thank you for youre response. A below image is the sample image.

sample

The assumption is needed because the program assumes position x, y, z values are written in world coordinates.

**Comment by sshige** [comment]
Thank you for your response.

Yes, the grasp poses candidates are visualized with the axes.

Sorry, "the point cloud is in the world coordinate", is misleading. Please transform input cloud to a flame which direction is the same as the robot is correct.

This is needed for satisfying the below function.
- direction: from which direction a robot try to grasp a target object. ('x': from the front, 'z': from the top)

For example, if the objects are on the table, we can use 'z' option to grasp. If the objects are in a shelf, we can use 'x' option to grasp.

**Comment by wkentaro** [comment]
Ok, I understood. Could you add a test and documentation with sample images?

**Comment by sshige** [comment]
Ok, I will do. 

I will let you know once I finish.

**Comment by wkentaro** [comment]
It looks good.

Anyway, are there any videos to demonstrate the feature of this node?
I and others may want to watch it.

**Comment by sshige** [comment]
Sorry for late.

The below one is a simple example.

out

**Comment by wkentaro** [comment]
Don’t you have the video by real camera?
And could you add the gif files to the doc?

**Comment by sshige** [comment]
Sorry for the very late response.

I don't have the video by real camera at now. I can take it next week.

I added the gif file on the doc.

File: jsk_pcl_ros/scripts/detect_graspable_poses_pcabase.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
+#!/usr/bin/env python
+
+from __future__ import print_function
+
+import numpy as np
+import matplotlib.pyplot as plt
```
Comment (knorth55) [inline jsk_pcl_ros/scripts/detect_graspable_poses_pcabase.py]: unnecessary import

File: jsk_perception/node_scripts/shigematsu_mask_rcnn_instance_segmentation.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env python
```
Comment (knorth55) [inline jsk_perception/node_scripts/shigematsu_mask_rcnn_instance_segmentation.py]: is this file needed?
if so, please rename it.

File: jsk_perception/sample/shigematsu_mask_rcnn_instance_segmentation.launch [hunk 3]
```diff
@@ -0,0 +1,183 @@
+<launch>
```
Comment (knorth55) [inline jsk_perception/sample/shigematsu_mask_rcnn_instance_segmentation.launch]: is this file needed?

File: jsk_pcl_ros/scripts/detect_graspable_poses_pcabase.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+        rospy.logwarn("This node is experiential one.")
+
+        super(DetectGraspablePosesPcabase, self).__init__()
+
+        self.direction = rospy.get_param('~direction', 'x')
+        self.hand_size = rospy.get_param('~hand_size', 0.13) # maximum hand width
+        self.interval_m = rospy.get_param('~interval_m', '0.04') # interval of possible grasp
+
+        self.pub_targetPoses = self.advertise("~output/can_grasp_poses", PoseArray, queue_size=1)
```
Comment (knorth55) [inline jsk_pcl_ros/scripts/detect_graspable_poses_pcabase.py]: for `python` variables, it is common to use snake case `pub_target_poses`.
all variables should be written in snake case. (class is camel case)

File: jsk_pcl_ros/scripts/detect_graspable_poses_pcabase.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        self.sub.unregister()
+
+    def callback(self, point_cloud):
+
+        points = np.array(list(pc2.read_points(point_cloud, skip_nans=True)))
+
+        interval_m = self.interval_m
+
+        if self.direction == 'z':
+            points_surface = points[:,[0,1]] # choose x, y
```
Comment (knorth55) [inline jsk_pcl_ros/scripts/detect_graspable_poses_pcabase.py]: please put spaces as `points_surface = points[:, [0, 1]]`

**Comment by sshige** [comment]
@knorth55 
Thank you for your kind reviews. I am sorry I uploaded two unnecessary files from lack of attention.

I think all places you pointed out are fixed. Could you check?

**Comment by knorth55** [comment]
I checked and all the codes are refactored. Thank you :)

File: jsk_pcl_ros/scripts/detect_graspable_poses_pcabase.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+        super(DetectGraspablePosesPcabase, self).__init__()
+
+        self.direction = rospy.get_param('~direction', 'x')
+        self.hand_size = rospy.get_param('~hand_size', 0.13) # maximum hand width
+        self.interval_m = rospy.get_param('~interval_m', '0.04') # interval of possible grasp
+
+        self.pub_target_poses = self.advertise("~output/can_grasp_poses", PoseArray, queue_size=1)
+
+        rospy.spin()
```
Comment (wkentaro) [inline jsk_pcl_ros/scripts/detect_graspable_poses_pcabase.py]: What about moving `rospy.init_node` and `rospy.spin` into the block of `if __name__ == '__main__'`?

File: doc/jsk_pcl_ros/nodes/detect_graspable_poses_pcabase.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
+# DetectGraspablePosesPcabase
```
Comment (wkentaro) [inline doc/jsk_pcl_ros/nodes/detect_graspable_poses_pcabase.md]: Could you add the private links of the video to the doc for future users in the lab?

**Comment by sshige** [comment]
@wkentaro 
Thank you for your kind reviews.
I modified the parts you mentioned.

File: jsk_pcl_ros/scripts/detect_graspable_poses_pcabase.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+            pose.orientation.z = quaternion[2]
+            pose.orientation.w = quaternion[3]
+            pub_msg.poses.append(pose)
+
+        self.pub_target_poses.publish(pub_msg)
+
+def main():
+    try:
+        DetectGraspablePosesPcabase()
+    except rospy.ROSInterruptException:
```
Comment (wkentaro) [inline jsk_pcl_ros/scripts/detect_graspable_poses_pcabase.py]: Is this try-except block necessary?
We usually don't have such block in jsk_perception.
https://github.com/jsk-ros-pkg/jsk_recognition/tree/master/jsk_perception/node_scripts

**Comment by wkentaro** [comment]
@sshige Sorry for the late response. I added a single comment above but it's a kind of coding style and other looks good.

**Comment by sshige** [comment]
@wkentaro 
Sorry for the delay. It's not necessary. I modified the part as you suggested.

**Comment by wkentaro** [comment]
Thanks! LGTM!
